A47-xds-federation.md

```diff
@@ -0,0 +1,682 @@
+of credentials that we grok from the bootstrap file).
+
+In the `XdsClient` API, the `CdsUpdate` struct will need to be changed to
+use an `XdsServer` struct instead of a simple string for the LRS server.
+The `XdsServer` struct will be populated with the name and channel creds
+of the xDS server from which the client obtained the CDS resource (i.e.,
+the `SelfConfigSource` will be expanded to an actual name, which can be
+passed down to the LB policies).  Note that this means that in the
+future, when we add support for xDS server fail-over, if we fail over to
+a different xDS server for a CDS resource containing a `SelfConfigSource`
+for the LRS field, the `XdsClient` will need to send a new update for the
+CDS resource to update the LRS server, even though the actual underlying
+resource did not change.
+
+The CDS LB policy will need to use this new field to populate the LB
+policy config for the `xds_cluster_resolver` LB policy.
+
+##### LB Policy Config Changes
+
+The configs for the `xds_cluster_resolver` and `xds_cluster_impl` LB
+policies will change to use type `XdsServer` instead of type
+`google.protobuf.StringValue` for the `lrs_load_reporting_server_name`
+field.
+
+The same `XdsServer::Parse()` method will be used in both the bootstrap
+file parser and the LB policy config parser in the `xds_cluster_resolver`
+and `xds_cluster_impl LB policies`.
+
+##### `XdsClient` Load Reporting APIs
+
+The load reporting APIs will need to accept the LRS server as an
+`XdsServer` struct instead of as a simple string.
+
+The load reporting APIs need to store data on a per-LRS-server basis.
+(At least in C-core, we are currently passing the LRS server name into
+the load reporting APIs, but we're not actually using it for anything
+internally.)  Note that in the future, if/when we implement xDS server
+failover, this means that when we fail over to a new server, we will
+wind up dropping a bit of data that was queued up for the old server,
+but that's probably fine.
+
+Note that we currently do not want to trust the control plane to tell us
+to contact any arbitrary server with GoogleDefaultCredentials, since
+this could leak the oauth token.  We also don't want that attack to
+happen via a service config distributed via a DNS TXT record that
+configures the `xds_cluster_impl` LB policy to send load reports to some
+arbitrary server with GoogleDefaultCredentials.  To avoid this, the
+`XdsClient` load reporting APIs will require that the LRS server passed
+into them is one of the servers that is already configured in the
+bootstrap file (i.e., it matches server name, channel creds, and server
+features).
+
+### Temporary environment variable protection
+
+While this feature is in development (until it has passed interop
+testing and is deemed production-ready), parsing of the new bootstrap
+config fields will be disabled unless the
+`GRPC_EXPERIMENTAL_XDS_FEDERATION` environment variable is set to true.
```

Comment:
Should we use `GRPC_XDS_EXPERIMENTAL_FEDERATION`, just to be consistent with the other env vars?  (swap `XDS` and `EXPERIMENTAL`)

e.g. RBAC is `GRPC_XDS_EXPERIMENTAL_RBAC` https://github.com/grpc/proposal/blob/master/A41-xds-rbac.md#temporary-environment-variable-protection

Reply:
Let's try to use the `GRPC_EXPERIMENTAL_` prefix going forward.  That way, we can have the same convention for both xDS and non-xDS functionality.